devDocs/startupShutdown.md

```diff
@@ -0,0 +1,178 @@
+
+1. No more than one NVDA process instance should be running time. Interactions with itself could cause severe issues, some (non-exhaustive list) examples of sub-systems where this would be a problem:
+   - NVDA config files
+   - Global (OS level) keyboard hook
+   - Changed / incompatible in-process code
+2. As such, we want to be able to detect running instances, cause them to exit, and confirm they have exited.
+
+### Exit hooks/triggers
+
+There are 3 ways that NVDA receives a request to exit:
+
+- From internally calling [triggerNVDAExit](#When-exiting-from-triggerNVDAExit)
+- Receiving [WM_QUIT](#When-exiting-from-WM_QUIT) Windows message
+- Receiving [wx.EVT_END_SESSION](#When-exiting-from-wxEVT_END_SESSION) due to Windows session ending
+
+### When exiting from `triggerNVDAExit`
+* Called from within NVDA.
+* A function in the core module
+* Only executes the code once, uses a lock and flag to ensure this
+* Uses a queue on the main thread to queue a safe shutdown
+* Once the queued shutdown starts:
+    1. the updateCheck is terminated
+    1. watchdog is terminated
+    1. addons and the brailleViewer are terminated, so we can close all windows safely
+    1. All wx windows are closed
+    1. Now that windows are closed, a new NVDA instance is started if requested
+
+### When exiting from `WM_QUIT`
+* [A Windows Message](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-quit) received from an external process, such as another NVDA process.
+* NVDA accepts `WM_QUIT` messages from other processes and creates a [named window](https://docs.microsoft.com/en-us/windows/win32/learnwin32/creating-a-window#creating-the-window) that can be discovered.
+* `WM_QUIT` is handled by `wx`, which force closes all wx windows (other UI features like the systray icon are not windows, and remain) and then exits the main loop.
+`triggerNVDAExit` is a more expansive check than how wxWidgets handles `WM_QUIT`
+* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.
+* Using a custom message has been considered:
+  - Would allow custom handling (eg just `triggerNVDAExit`)
+  - Unfortunately, older NVDA versions will only be aware of `WM_QUIT`, so we'd need to send `WM_QUIT` to these versions.
+  - Sending the custom message, waiting for a timeout, then sending `WM_QUIT` adds a significant wait time
+  - Identifying the running version (to selectively send the message) requires maintaining 2 message windows in NVDA (one for legacy behaviour) and adds complexity
+
+### When exiting from `wx.EVT_END_SESSION`
+* This is a [wxCloseEvent](https://docs.wxwidgets.org/3.0/classwx_close_event.html) triggered by a Windows session ending.
+* On `wx.EVT_END_SESSION`, we save the config and play the exit sound.
+* Other actions are not performed as we have limited time to perform an action for this event.
+NVDA is expected to run as long as possible during the sign out process.

```

Comment:
```suggestion
NVDA is expected to run as long as possible during the sign out process.
This is achieved through the Windows API, using [SHUTDOWN_NORETRY](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessshutdownparameters)
```

Reply:
I think the important part (for shutdown order) is the `dwLevel` parameter: 

> The shutdown priority for a process relative to other processes in the system. The system shuts down processes from high dwLevel values to low. 

> dwLevel of `100-1FF` application reserved last shutdown range.

Reply:
I think it would be worth pointing that out.

Reply:
Perhaps even note that value 0x100 is used as the lowest value reserved for non system applications